Hw11 solid

Taisiya-Boon opened the pull request:
hw11 Написать эмулятор АТМ (банкомата). 

Объект класса АТМ должен уметь:
     - принимать банкноты разных номиналов (на каждый номинал должна быть своя ячейка).
     - выдавать запрошенную сумму минимальным количеством банкнот или ошибку если сумму нельзя выдать.
     - выдавать сумму остатка денежных средств.

Taisiya-Boon committed:
hw11 Написать эмулятор АТМ (банкомата).

     Объект класса АТМ должен уметь:
     - принимать банкноты разных номиналов (на каждый номинал должна быть своя ячейка).
     - выдавать запрошенную сумму минимальным количеством банкнот или ошибку если сумму нельзя выдать.
     - выдавать сумму остатка денежных средств.

Taisiya-Boon committed:
hw11 добавление изменений в корневом pom-файле

kataus commented on hw11-solid/src/main/java/ru/otus/ATM.java:
```diff
@@ -0,0 +1,136 @@
+
+public class ATM {
+
+    private Cell cellNom10 = new Cell(10);
```
Банкноты лучше сделать enum или хотя бы списком доступных. Чтобы оперировать не по номиналу, а с массивом данных. Иначе любое изменение банкнотного состава станет нашим Ватерло

Taisiya-Boon committed:
hw11 добавление интерфейса для ячеек банкомата;
     добавление enum для номиналов;
     переработка некоторых методов класса ATM

Taisiya-Boon committed:
hw11 добавление интерфейса для ячеек банкомата;
     добавление enum для номиналов;
     переработка некоторых методов класса ATM

kataus commented on hw11-solid/src/main/java/ru/otus/CellImpl.java:
```diff
@@ -0,0 +1,68 @@
+        return parBanknote;
+    }
+
+    private enum ParBanknote {
```
А почему приватный? Как мы будем оперировать номиналами, если о них никто не знает?

kataus commented on hw11-solid/src/main/java/ru/otus/CellImpl.java:
```diff
@@ -0,0 +1,68 @@
+
+    CellImpl(int parBanknote){
+        int bufferPar = 0;
+        for (ParBanknote par : ParBanknote.values()) {
```
Задача ячейки принимать и давать банковские листы, но никак не знать об их покупательной способности

kataus commented on hw11-solid/src/main/java/ru/otus/ATM.java:
```diff
@@ -0,0 +1,92 @@
+
+public class ATM {
+
+    private CellImpl cellNom10 = new CellImpl(10);
```
А если у нас введут новую купюру? Придётся всю систему вручную проверять чтобы было корректно? Давайте таки enum сделаем не тайным знанием ячейки, а общедоступным справочником.

kataus commented on hw11-solid/src/main/java/ru/otus/ATM.java:
```diff
@@ -0,0 +1,92 @@
+package ru.otus;
+
+public class ATM {
```
А у  ATM интерфейс почему решили не выделять?

Taisiya-Boon committed:
hw11 добавление интерфейса для ATM;
     выделение отдельного файла под enum для номиналов;
     переработка конструктора класса ATM

kataus commented on hw11-solid/src/main/java/ru/otus/ATM.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface ATM {
+
+    public int [] addMoney(int... money);
```
Не надо использовать массивы без особой необходимости. Коллекции гораздо удобнее и безопаснее

kataus commented on hw11-solid/src/main/java/ru/otus/CellImpl.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+class CellImpl implements Cell {
+
+    private final int parBanknote;
```
ЛУчше тут напрямую ENUM указывать, так будет удобнее гораздо везде. И очевиднее

Taisiya-Boon committed:
hw11 в классе CellImpl изменён тип ячейки parBanknote на enum;
     в классе ATMImpl изменены типы возвращаемого значения методов addMoney и takeMoney на ArrayList

kataus commented on hw11-solid/src/main/java/ru/otus/ATM.java:
```diff
@@ -1,10 +1,12 @@
 public interface ATM {
 
-    public int [] addMoney(int... money);
+    public ArrayList addMoney(int... money);
```
корректно возвращать описание, а не зажимать в конкретную реализацию. Для нас нет разницы какая имлементация List вернется.

Taisiya-Boon committed:
hw11 в интерфейсе ATM изменнён тип возвращаемого значения мотодов addMoney и takeMoney, с узконаправленного ArrayList на более широкий Collection
